rust/projects/parallelism/src/engines/kvs.rs

```diff
@@ -0,0 +1,462 @@
+/// assert_eq!(val, Some("value".to_owned()));
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct KvStore {
+    // directory for the log and other data
+    path: Arc<PathBuf>,
+    // map generation number to the file reader
+    reader: KvStoreReader,
+    index: Arc<SkipMap<String, CommandPos>>,
+    writer: Arc<Mutex<KvStoreWriter>>,
+}
+
+/// A single thread reader.
+///
+/// Each `KvStore` instance has its own `KvStoreReader` and
+/// `KvStoreReader`s open the same files separately. So the user
+/// can read concurrently through multiple `KvStore`s in different
+/// threads.
+struct KvStoreReader {
+    path: Arc<PathBuf>,
+    // generation of the latest compaction file
+    // readers with a generation before safe_point can be closed
+    safe_point: Arc<AtomicU64>,
+    readers: RefCell<BTreeMap<u64, BufReaderWithPos<File>>>,
+}
+
+impl KvStoreReader {
+    /// Read the log file at the given `CommandPos`.
+    fn read_and<F, R>(&self, cmd_pos: CommandPos, f: F) -> Result<R>
+    where
+        F: FnOnce(io::Take<&mut BufReaderWithPos<File>>) -> Result<R>,
+    {
+        let mut readers = self.readers.borrow_mut();
+
+        // Close readers of stale files
+        while !readers.is_empty() {
+            let first_gen = *readers.keys().next().unwrap();
+            if self.safe_point.load(Ordering::SeqCst) <= first_gen {
+                break;
+            }
+            readers.remove(&first_gen);
+        }
+
+        // Open the file if we haven't opened it in this `KvStoreReader`.
+        // We don't use entry API here because we want the errors to be propogated.
+        if !readers.contains_key(&cmd_pos.gen) {
+            let reader = BufReaderWithPos::new(File::open(log_path(&self.path, cmd_pos.gen))?)?;
+            readers.insert(cmd_pos.gen, reader);
+        }
+        let reader = readers.get_mut(&cmd_pos.gen).unwrap();
+        reader.seek(SeekFrom::Start(cmd_pos.pos))?;
+        let cmd_reader = reader.take(cmd_pos.len);
+        f(cmd_reader)
+    }
+
+    // Read the log file at the given `CommandPos` and deserialize it to `Command`.
+    fn read_command(&self, cmd_pos: CommandPos) -> Result<Command> {
+        self.read_and(cmd_pos, |cmd_reader| {
+            Ok(serde_json::from_reader(cmd_reader)?)
+        })
+    }
+}
+
+impl Clone for KvStoreReader {
+    fn clone(&self) -> KvStoreReader {
+        KvStoreReader {
+            path: Arc::clone(&self.path),
+            safe_point: Arc::clone(&self.safe_point),
+            // don't use other KvStoreReader's readers
+            readers: RefCell::new(BTreeMap::new()),
+        }
+    }
+}
+
+struct KvStoreWriter {
+    reader: KvStoreReader,
+    writer: BufWriterWithPos<File>,
+    current_gen: u64,
+    // the number of bytes representing "stale" commands that could be
+    // deleted during a compaction
+    uncompacted: u64,
+    path: Arc<PathBuf>,
+    index: Arc<SkipMap<String, CommandPos>>,
+}
+
+impl KvStore {
```

Comment:
The ordering of items here is inconsistent: KvStore struct, KvStoreReader struct, KvStoreReader impls, KvStoreWriter struct, KvStore impls, KvStoreWriter impls.

Can you either define all the structs together, then the impls in the same order; or define each struct's impls immediately after their def?